Make the table name too long error message more explicit for Postgres

adampal opened the pull request:
Fixes #49557

Makes it explicit why the enforced limit for max table name length is less than the default Postgres limit.

adampal committed:
Make the table name too long error message more explicit for PostgreSQL

* Explain why the enforced limit is less than the Postgres limit

Extract the table name too long error message out into its own method

Fix tests for new table name too long error message

Update error message from Postgres to PostgreSQL

ghiculescu commented:
You'll need to update all the tests added in https://github.com/rails/rails/pull/45136. I also think maybe it's better to extract the error message out to its own method so that the check logic isn't duplicated between adapters.

adampal commented:
Thanks @ghiculescu I've moved the error message out into its own method and fixed the tests.
Are you happy with the `if current_adapter?(:PostgreSQLAdapter)` statement in the tests?

ghiculescu commented:
Once the tests are green, could you squash your commits please.

yahonda commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -608,6 +608,10 @@ def table_name_length
       end
 
+      def table_name_too_long_error_message(table_name)
+        "Table name '#{table_name}' is too long; Postgres allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that Postgres adds to table names when creating default indexes, so our effective limit is #{table_name_length} characters"
```
```suggestion
        "Table name '#{table_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{table_name_length} characters"
```

yahonda commented on activerecord/test/cases/migration/rename_table_test.rb:
```diff
@@ -52,13 +52,18 @@ def test_rename_table
         end
-        assert_equal "Table name '#{long_name}' is too long; the limit is #{name_limit} characters", error.message
+        if current_adapter?(:PostgreSQLAdapter)
+          assert_equal "Table name '#{long_name}' is too long; Postgres allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that Postgres adds to table names when creating default indexes, so our effective limit is #{name_limit} characters", error.message
```
```suggestion
          assert_equal "Table name '#{long_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{name_limit} characters", error.message
```

yahonda commented on activerecord/test/cases/migration_test.rb:
```diff
@@ -200,13 +200,18 @@ def test_create_table_with_if_not_exists_true
     end
-    assert_equal "Table name '#{long_name}' is too long; the limit is #{name_limit} characters", error.message
+    if current_adapter?(:PostgreSQLAdapter)
+      assert_equal "Table name '#{long_name}' is too long; Postgres allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that Postgres adds to table names when creating default indexes, so our effective limit is #{name_limit} characters", error.message
```
```suggestion
      assert_equal "Table name '#{long_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{name_limit} characters", error.message
```

adampal commented:
Thanks very much @yahonda and @ghiculescu 
I've made the changes and squashed the commits.
Let me know if there's anything else you think needs to happen.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1773,10 +1773,14 @@ def validate_index_length!(table_name, new_name, internal = false)
         end
 
+        def table_name_too_long_error_message(table_name)
+          "Table name '#{table_name}' is too long; the limit is #{table_name_length} characters"
```
We usually end errors with periods. I noticed this was like that before, but let's add one now.

```suggestion
          "Table name '#{table_name}' is too long; the limit is #{table_name_length} characters."
```

adampal replied:
Thanks @eileencodes I've added them.

eileencodes commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -608,6 +608,10 @@ def table_name_length
       end
 
+      def table_name_too_long_error_message(table_name)
+        "Table name '#{table_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{table_name_length} characters"
```
```suggestion
        "Table name '#{table_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{table_name_length} characters."
```

matthewd commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -608,6 +608,10 @@ def table_name_length
       end
 
+      def table_name_too_long_error_message(table_name)
+        "Table name '#{table_name}' is too long; PostgreSQL allows a maximum length of #{max_identifier_length} characters, but we also need to allow for the `_pkey` suffix that PostgreSQL adds to table names when creating default indexes, so our effective limit is #{table_name_length} characters"
```
While this message is _accurate_, it feels like a rather unergonomic info-dump instead of a short actionable error message.

While I can appreciate the confusion that our enforced limit differs from PostgreSQL's, I still don't feel like this is the place to tell that story.

adampal replied:
@matthewd agreed, it's a brain dump of an error message.  This feels like a short term fix to me. I don't think we should be limiting table names to 58 characters given it's not actually a Postgres limitation. The issue is actually with index names. A table name of 63 characters will work fine so I think we should allow it and instead fix the index names if they are too long.

I didn't want to pollute this PR and was going to submit a separate issue to allow table names of 63 characters again.

I thought improving the error message might be something we could do quickly to help people upgrading to Rails 7.1 as it caused me a bit of confusion.

Depending on what everyone else thinks, I'm happy to submit a separate PR to allow table names up to 63 characters again.

p8 commented:
Would it be possible to automatically shorten the table name? Similar to https://github.com/rails/rails/pull/47753

matthewd commented:
@p8 I don't think so -- index names are mostly unimportant after definition, whereas expectations and mentions of table names appear all over the place.

ghiculescu commented:
From memory the original issue was that dropping tables with long names doesn’t work because the index names matter then. I suppose we could fix that more directly?

rafaelfranca commented:
Does #49592 fix the issue?

adampal commented:
> Does #49592 fix the issue?

Yes, that's a much better solution. I'll close this PR.
